hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieMergeOnReadSnapshotReader.java

```diff
@@ -0,0 +1,192 @@
+import static org.apache.hudi.common.config.HoodieCommonConfig.SPILLABLE_DISK_MAP_TYPE;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.COMPACTION_LAZY_BLOCK_READ_ENABLED_PROP;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.DEFAULT_COMPACTION_LAZY_BLOCK_READ_ENABLED;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.DEFAULT_MAX_DFS_STREAM_BUFFER_SIZE;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.DEFAULT_SPILLABLE_MAP_BASE_PATH;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.ENABLE_OPTIMIZED_LOG_BLOCKS_SCAN;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.MAX_DFS_STREAM_BUFFER_SIZE_PROP;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.SPILLABLE_MAP_BASE_PATH_PROP;
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getBaseFileReader;
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getMaxCompactionMemoryInBytes;
+import static org.apache.hudi.internal.schema.InternalSchema.getEmptyInternalSchema;
+
+public class HoodieMergeOnReadSnapshotReader extends AbstractRealtimeRecordReader implements Iterator<HoodieRecord>, AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieMergeOnReadSnapshotReader.class);
+
+  private final String tableBasePath;
+  private final List<HoodieLogFile> logFilePaths;
+  private final String latestInstantTime;
+  private final Schema readerSchema;
+  private final JobConf jobConf;
+  private final HoodieMergedLogRecordScanner logRecordScanner;
+  private final HoodieFileReader baseFileReader;
+  private final Map<String, HoodieRecord> logRecordsByKey;
+  private final Iterator<HoodieRecord> recordsIterator;
+  private final ExternalSpillableMap<String, HoodieRecord> mergedRecordsByKey;
+
+  public HoodieMergeOnReadSnapshotReader(String tableBasePath, String baseFilePath,
+                                         List<HoodieLogFile> logFilePaths,
+                                         String latestInstantTime,
+                                         Schema readerSchema,
+                                         JobConf jobConf, long start, long length, String[] hosts) throws IOException {
+    super(getRealtimeSplit(tableBasePath, baseFilePath, logFilePaths, latestInstantTime, start, length, hosts), jobConf);
+    this.tableBasePath = tableBasePath;
+    this.logFilePaths = logFilePaths;
+    this.latestInstantTime = latestInstantTime;
+    this.readerSchema = readerSchema;
+    this.jobConf = jobConf;
+    HoodieTimer timer = new HoodieTimer().startTimer();
+    this.logRecordScanner = getMergedLogRecordScanner();
+    LOG.debug("Time taken to scan log records: {}", timer.endTimer());
+    this.baseFileReader = getBaseFileReader(new Path(baseFilePath), jobConf);
+    this.logRecordsByKey = logRecordScanner.getRecords();
```

Comment:
Do we need an external spillable map for the log records as well?

Reply:
external spillable map is used to hold merged records by key. `logRecordsByKey` is a simple map. We can do with simple map for merged records too but I wasn't sure how many records can there be across base and log files and what resources user env has, so kept external spillable map.

Reply:
I was thinking if the size of records in `logRecordsByKey` can exceed the memory and why we use a simple map for `logRecordsByKey` but not `mergedRecordsByKey`, which is my concern.  If we use the simple map as well in the existing realtime reader it should be OK.  Is that the case?